### PR TITLE
binwalk, diffoscope: update after Python 3.12

### PR DESCRIPTION
**Pull request by philiptaron**
## Description of changes

- [x] `binwalk` uses both `imp` and `nose` which are not supported on Python 3.12. See https://github.com/ReFirmLabs/binwalk/pull/668
- [x] `diffoscope` unit test for binwalk now sees more cpio binaries in its test.

related to https://github.com/NixOS/nixpkgs/issues/325657#issuecomment-2215424154
cc @Mag1cByt3s

Result of `nixpkgs-review` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>15 packages built:</summary>
  <ul>
    <li>binwalk (python312Packages.binwalk)</li>
    <li>binwalk.dist (python312Packages.binwalk.dist)</li>
    <li>diffoscope</li>
    <li>diffoscope.dist</li>
    <li>diffoscope.man</li>
    <li>diffoscopeMinimal</li>
    <li>diffoscopeMinimal.dist</li>
    <li>diffoscopeMinimal.man</li>
    <li>ghidra-extensions.ghidraninja-ghidra-scripts</li>
    <li>python311Packages.binwalk</li>
    <li>python311Packages.binwalk-full</li>
    <li>python311Packages.binwalk-full.dist</li>
    <li>python311Packages.binwalk.dist</li>
    <li>python312Packages.binwalk-full</li>
    <li>python312Packages.binwalk-full.dist</li>
  </ul>
</details>

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by philiptaron** [comment]
@trofi fyi

**Comment by emilazy** [comment]
Is that behavioural change in `binwalk` benign? I’m curious as to why a Python upgrade and an import internals change would lead to it finding more binaries.

**Comment by philiptaron** [comment]
> Is that behavioural change in binwalk benign? I’m curious as to why a Python upgrade and an import internals change would lead to it finding more binaries.

I was quite surprised to see it as well. You can see from the patch what the detected differences are. 

I checked out upstream (https://salsa.debian.org/reproducible-builds/diffoscope) locally and ran `master`'s copy of binwalk (for me, `/nix/store/ckv7akdj4cjl8qkiij8vraiisw9n6mvj-python3.11-binwalk-2.3.4`):

```
$ /nix/store/ckv7akdj4cjl8qkiij8vraiisw9n6mvj-python3.11-binwalk-2.3.4/bin/binwalk tests/data/test{1,2}.binwalk

Scan Time:     2024-07-08 13:07:08
Target File:   /home/philip/Code/salsa.debian.org/reproducible-builds/diffoscope/tests/data/test1.binwalk
MD5 Checksum:  03bb6be017bd9db1586ee5cc291dd992
Signatures:    411

DECIMAL       HEXADECIMAL     DESCRIPTION
--------------------------------------------------------------------------------
0             0x0             ASCII cpio archive (SVR4 with no CRC), file name: ".", file name length: "0x00000002", file size: "0x00000000"
112           0x70            ASCII cpio archive (SVR4 with no CRC), file name: "dir", file name length: "0x00000004", file size: "0x00000000"
228           0xE4            ASCII cpio archive (SVR4 with no CRC), file name: "dir/link", file name length: "0x00000009", file size: "0x00000006"
356           0x164           ASCII cpio archive (SVR4 with no CRC), file name: "dir/text", file name length: "0x00000009", file size: "0x000001BE"
924           0x39C           ASCII cpio archive (SVR4 with no CRC), file name: "TRAILER!!!", file name length: "0x0000000B", file size: "0x00000000"
1536          0x600           ASCII cpio archive (SVR4 with no CRC), file name: ".", file name length: "0x00000002", file size: "0x00000000"
1648          0x670           ASCII cpio archive (SVR4 with no CRC), file name: "dir", file name length: "0x00000004", file size: "0x00000000"
1764          0x6E4           ASCII cpio archive (SVR4 with no CRC), file name: "dir/link", file name length: "0x00000009", file size: "0x0000000D"
1900          0x76C           ASCII cpio archive (SVR4 with no CRC), file name: "dir/text", file name length: "0x00000009", file size: "0x0000029F"
2692          0xA84           ASCII cpio archive (SVR4 with no CRC), file name: "TRAILER!!!", file name length: "0x0000000B", file size: "0x00000000"


Scan Time:     2024-07-08 13:07:08
Target File:   /home/philip/Code/salsa.debian.org/reproducible-builds/diffoscope/tests/data/test2.binwalk
MD5 Checksum:  d47335a3e695dc2b3dd8bfc56ba476a5
Signatures:    411

DECIMAL       HEXADECIMAL     DESCRIPTION
--------------------------------------------------------------------------------
0             0x0             ASCII cpio archive (SVR4 with no CRC), file name: ".", file name length: "0x00000002", file size: "0x00000000"
112           0x70            ASCII cpio archive (SVR4 with no CRC), file name: "dir", file name length: "0x00000004", file size: "0x00000000"
228           0xE4            ASCII cpio archive (SVR4 with no CRC), file name: "dir/link", file name length: "0x00000009", file size: "0x0000000D"
364           0x16C           ASCII cpio archive (SVR4 with no CRC), file name: "dir/text", file name length: "0x00000009", file size: "0x0000029F"
1156          0x484           ASCII cpio archive (SVR4 with no CRC), file name: "TRAILER!!!", file name length: "0x0000000B", file size: "0x00000000"
1536          0x600           ASCII cpio archive (SVR4 with no CRC), file name: ".", file name length: "0x00000002", file size: "0x00000000"
1648          0x670           ASCII cpio archive (SVR4 with no CRC), file name: "dir", file name length: "0x00000004", file size: "0x00000000"
1764          0x6E4           ASCII cpio archive (SVR4 with no CRC), file name: "dir/link", file name length: "0x00000009", file size: "0x00000006"
1892          0x764           ASCII cpio archive (SVR4 with no CRC), file name: "dir/text", file name length: "0x00000009", file size: "0x000001BE"
2460          0x99C           ASCII cpio archive (SVR4 with no CRC), file name: "TRAILER!!!", file name length: "0x0000000B", file size: "0x00000000"
```

Then, with this PR (note the Python 3.12 in the derivation name):

```
$ /nix/store/2c7kkgy4hbvxjh1rj0cvcvc0r1yb136l-python3.12-binwalk-2.3.4/bin/binwalk tests/data/test{1,2}.binwalk

Scan Time:     2024-07-08 13:08:08
Target File:   /home/philip/Code/salsa.debian.org/reproducible-builds/diffoscope/tests/data/test1.binwalk
MD5 Checksum:  03bb6be017bd9db1586ee5cc291dd992
Signatures:    411

DECIMAL       HEXADECIMAL     DESCRIPTION
--------------------------------------------------------------------------------
0             0x0             ASCII cpio archive (SVR4 with no CRC), file name: ".", file name length: "0x00000002", file size: "0x00000000"
112           0x70            ASCII cpio archive (SVR4 with no CRC), file name: "dir", file name length: "0x00000004", file size: "0x00000000"
228           0xE4            ASCII cpio archive (SVR4 with no CRC), file name: "dir/link", file name length: "0x00000009", file size: "0x00000006"
356           0x164           ASCII cpio archive (SVR4 with no CRC), file name: "dir/text", file name length: "0x00000009", file size: "0x000001BE"
924           0x39C           ASCII cpio archive (SVR4 with no CRC), file name: "TRAILER!!!", file name length: "0x0000000B", file size: "0x00000000"
1536          0x600           ASCII cpio archive (SVR4 with no CRC), file name: ".", file name length: "0x00000002", file size: "0x00000000"
1648          0x670           ASCII cpio archive (SVR4 with no CRC), file name: "dir", file name length: "0x00000004", file size: "0x00000000"
1764          0x6E4           ASCII cpio archive (SVR4 with no CRC), file name: "dir/link", file name length: "0x00000009", file size: "0x0000000D"
1900          0x76C           ASCII cpio archive (SVR4 with no CRC), file name: "dir/text", file name length: "0x00000009", file size: "0x0000029F"
2692          0xA84           ASCII cpio archive (SVR4 with no CRC), file name: "TRAILER!!!", file name length: "0x0000000B", file size: "0x00000000"


Scan Time:     2024-07-08 13:08:08
Target File:   /home/philip/Code/salsa.debian.org/reproducible-builds/diffoscope/tests/data/test2.binwalk
MD5 Checksum:  d47335a3e695dc2b3dd8bfc56ba476a5
Signatures:    411

DECIMAL       HEXADECIMAL     DESCRIPTION
--------------------------------------------------------------------------------
0             0x0             ASCII cpio archive (SVR4 with no CRC), file name: ".", file name length: "0x00000002", file size: "0x00000000"
112           0x70            ASCII cpio archive (SVR4 with no CRC), file name: "dir", file name length: "0x00000004", file size: "0x00000000"
228           0xE4            ASCII cpio archive (SVR4 with no CRC), file name: "dir/link", file name length: "0x00000009", file size: "0x0000000D"
364           0x16C           ASCII cpio archive (SVR4 with no CRC), file name: "dir/text", file name length: "0x00000009", file size: "0x0000029F"
1156          0x484           ASCII cpio archive (SVR4 with no CRC), file name: "TRAILER!!!", file name length: "0x0000000B", file size: "0x00000000"
1536          0x600           ASCII cpio archive (SVR4 with no CRC), file name: ".", file name length: "0x00000002", file size: "0x00000000"
1648          0x670           ASCII cpio archive (SVR4 with no CRC), file name: "dir", file name length: "0x00000004", file size: "0x00000000"
1764          0x6E4           ASCII cpio archive (SVR4 with no CRC), file name: "dir/link", file name length: "0x00000009", file size: "0x00000006"
1892          0x764           ASCII cpio archive (SVR4 with no CRC), file name: "dir/text", file name length: "0x00000009", file size: "0x000001BE"
2460          0x99C           ASCII cpio archive (SVR4 with no CRC), file name: "TRAILER!!!", file name length: "0x0000000B", file size: "0x00000000"
```

What I don't understand is how there were _ever_ two emitted items. Master appears to have 10 as well.

File: pkgs/tools/misc/diffoscope/binwalk-python3.12.patch [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+diff --git a/tests/comparators/test_binwalk.py b/tests/comparators/test_binwalk.py
+index ba0b1bd4..c85b5934 100644
+--- a/tests/comparators/test_binwalk.py
++++ b/tests/comparators/test_binwalk.py
+@@ -57,7 +57,7 @@ def comparison(binwalk1, binwalk2):
+ @skip_unless_module_exists("binwalk")
+ def test_listing(comparison):
+     differences = comparison.details
+-    assert comparison.comments == ["comprises of 2 embedded members"]
++    assert comparison.comments == ["comprises of 10 embedded members"]
```
Comment (eclairevoyant) [inline pkgs/tools/misc/diffoscope/binwalk-python3.12.patch:35]: is this by design i.e. a number that we can rely on to continue to be 10? or might this change in the future?

**Comment by superherointj** [comment]
Please rebase.

**Comment by eclairevoyant** [comment]
Both binwalk and diffoscope appear to build on current `master`

**Comment by superherointj** [comment]
> Both binwalk and diffoscope appear to build on current `master` ([6bf4ef3](https://github.com/NixOS/nixpkgs/commit/6bf4ef3fafa1746608e08a21709e3f77f0790e8c))

Yes, because I merged #326295.

**Comment by emilazy** [comment]
But there’s nothing to rebase now; #326295 replaces this.

**Comment by eclairevoyant** [comment]
Right, what I'm saying is I believe this PR can be closed as it appears to be a build fix. Unless there was some other plan with the PR as well.

**Comment by philiptaron** [comment]
> But there’s nothing to rebase now; #326295 replaces this.

Quite so.

**Pull request closed without merging**